src/sage/rings/polynomial/infinite_polynomial_element.py

```diff
@@ -578,9 +616,70 @@ def variables(self):
 
         """
         if hasattr(self._p, 'variables'):
-            return tuple(self._p.variables())
+            P = self.parent()
+            return tuple(InfinitePolynomial(P, v) for v in self._p.variables())
         return ()
 
+    def monomials(self):
+        """
+        Return the list of monomials in self. The returned list is
+        decreasingly ordered by the term ordering of
+        ``self.parent()``.
+
+        EXAMPLES::
+
+            sage: X.<x> = InfinitePolynomialRing(QQ)
+            sage: p = x[1]^3 + x[2] - 2*x[1]*x[3]
+            sage: p.monomials()
+            [x_3*x_1, x_2, x_1^3]
+
+            sage: X.<x> = InfinitePolynomialRing(QQ, order='deglex')
+            sage: p = x[1]^3 + x[2] - 2*x[1]*x[3]
+            sage: p.monomials()
+            [x_1^3, x_3*x_1, x_2]
+        """
+        P = self.parent()
+        return [InfinitePolynomial(P, m) for m in self._p.monomials()]
+
+    def monomial_coefficient(self, mon):
```

Comment:
Are you sure we need this method? Shouldn't the default redirection in`__getattr__` handle this?

Reply:
It seems so.  Without the method `monomial_coefficient` I have:
```
sage: X.<x> = InfinitePolynomialRing(QQ)
sage: f = 2*x[0]*x[2] + 3*x[1]^2
sage: c = f.monomial_coefficient(x[1]^2); c
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[4], line 1
----> 1 c = f.monomial_coefficient(x[Integer(1)]**Integer(2)); c

TypeError: Argument 'mon' has incorrect type (expected sage.rings.polynomial.multi_polynomial_libsingular.MPolynomial_libsingular, got InfinitePolynomial_dense)
```

Reply:
Ah, right, yes, we need to extract out the underlying polynomial.